src/main/java/com/redhat/devtools/lsp4ij/features/documentation/LSPDocumentationHelper.java

```diff
@@ -26,63 +30,115 @@ private LSPDocumentationHelper() {
         }
-        String htmlBody = MarkupKind.MARKDOWN.equals(content.getKind()) ?
-                MarkdownConverter.toHTML(content.getValue()) :
-                content.getValue();
-        return styleHtml(editor, htmlBody);
+        var documentation = completionItem.getDocumentation();
+        if (documentation == null) {
+            return Collections.emptyList();
+        }
+        if (documentation.isLeft()) {
+            String value = documentation.getLeft();
+            if (!isValidContent(value)) {
+                return Collections.emptyList();
+            }
+            return List.of(new MarkupContent(MarkupKind.PLAINTEXT, value));
+        } else if (documentation.isRight()) {
+            MarkupContent content = documentation.getRight();
+            if (!isValidContent(content)) {
+                return null;
+            }
+            return List.of(content);
+        }
+        return Collections.emptyList();
     }
 
-    private static String styleHtml(@Nullable Editor editor, String htmlBody) {
-        if (htmlBody == null || htmlBody.isEmpty()) {
-            return htmlBody;
+    public static @NotNull List<MarkupContent> getValidMarkupContents(@Nullable Hover hover) {
+        if (hover == null) {
+            // textDocument/hover may return null
+            return Collections.emptyList();
         }
-        Color background = editor != null ? editor.getColorsScheme().getDefaultBackground() : null;
-        Color foreground = editor != null ? editor.getColorsScheme().getDefaultForeground() : null;
-
-        StringBuilder html = new StringBuilder("<html><head><style TYPE='text/css'>html { ");
-        if (background != null) {
-            html.append("background-color: ")
-                    .append(toHTML(background))
-                    .append(";");
+        var hoverContents = hover.getContents();
+        if (hoverContents == null) {
+            return Collections.emptyList();
         }
-        if (foreground != null) {
-            html.append("color: ")
-                    .append(toHTML(foreground))
-                    .append(";");
+        if (hoverContents.isLeft()) {
+            var contents = hoverContents.getLeft();
+            if (contents == null || contents.isEmpty()) {
+                return Collections.emptyList();
+            }
+            return contents.stream()
+                    .map(content -> {
+                        if (content.isLeft()) {
+                            String value = content.getLeft();
+                            if (!isValidContent(value)) {
+                                return null;
+                            }
+                            return new MarkupContent(MarkupKind.MARKDOWN, value);
+                        } else if (content.isRight()) {
+                            MarkedString markedString = content.getRight();
+                            if (!isValidContent(markedString)) {
+                                return null;

```

Comment:
return should be  `@NotNull`

Reply:
It should, because it is filtered after.